src/black-magic.js

```diff
@@ -0,0 +1,371 @@
+  const mutateMethods = [
+    {
+      klass: Node,
+      methods: [
+        'insertBefore',
+        'appendChild',
+        'replaceChild',
+        'removeChild',
+      ],
+    },
+    {
+      klass: Element,
+      methods: [
+        'setAttribute',
+        'removeAttribute',
+      ],
+    },
+  ];
+  mutateMethods.forEach(({klass, methods}) => {
+    const proto = klass.prototype;
+
+    methods.forEach(prop => {
+      const descriptor = Object.getOwnPropertyDescriptor(proto, prop);
+
+      // We only care if this is a data descriptor (which guarantees the
+      // `value` is the function).
+      if (!descriptor || !descriptor.value) {
+        return;
+      }
+
+      const method = descriptor.value;
+      descriptor.value = function() {
+        assertMutationPhase(this);
+        return method.apply(this, arguments);
+      };
+      Object.defineProperty(descriptor.value, 'name', {
+        value: `${klass.name}#${prop}`,
+      });
+
+      Object.defineProperty(proto, prop, descriptor);
+    });
+  });
+
+  // Next, we'll monkey patch live-connected objects which can cause mutations
+  const liveConnectedMutators = [
+    {
+      // NamedNodeMap
+      klasses: [
+        {klass: Element, prop: 'attributes'},
+      ],
+      methods: [
+        'setNamedItem',
+        'removeNamedItem',
+      ],
+      setters: [],
+    },
+    {
+      // DOMTokenList
+      klasses: [
+        {klass: Element, prop: 'classList'},
+        {klass: window.HTMLLinkElement, prop: 'relList'},
+        {klass: window.HTMLAnchorElement, prop: 'relList'},
+        {klass: window.HTMLIFrameElement, prop: 'sandbox'},
+        {klass: window.HTMLAreaElement, prop: 'relList'},
+        {klass: window.HTMLOutputElement, prop: 'htmlFor'},
+      ],
+      methods: [
+        'add',
+        'remove',
+        'replace',
+        'toggle',
+
+      ],
+      setters: [
+        'value',
+      ],
+    },
+    {
+      // CSSStyleDeclaration
+      klasses: [
+        {klass: HTMLElement, prop: 'style'},
+      ],
+      methods: [
+        'setProperty',
+        'setPropertyValue',
+        'setPropertyPriority',
+        'removeProperty',
+      ],
+      setters: [
+        'cssText',
+        'cssFloat',
+      ],
+    },
+  ];
+  liveConnectedMutators.forEach(({klasses, methods, setters}) => {
+    klasses.forEach(({klass, prop}) => {
+      if (!klass) {
+        return;
+      }
+      const proto = klass.prototype;
+      const descriptor = Object.getOwnPropertyDescriptor(proto, prop);
+
+      // We only care if this is an accessor descriptor with a getter.
+      if (!descriptor || !descriptor.get) {
+        return;
+      }
+
+      const getter = descriptor.get;
+      descriptor.get = function() {
+        const element = this;
+        const liveObject = getter.call(this);
+
+        methods.forEach(prop => {
+          // Only install once
+          if (hasOwn(liveObject, prop) || !(prop in liveObject)) {
```

Comment:
How could this get installed twice?

Reply:
Every time you call `element.style`, this out function (the `descriptor.get`) gets called. So, if I do `element.style` 5 times, it would try to monkey patch all 5 times (which would get these really weird recursive looking stacks).

Reply:
I see, thanks.